### PR TITLE
Thai tone marks and vowels at the end of the search term is wrongly removed.

### PR DESCRIPTION
**Pull request by callmeott**
#### Summary

Thai vowels and tone marks at the end of the search term are now (likely unintentionally) removed.

E.g., สวัสดี will become only สวัสด, which is wrong.

This fix allows all Thai characters to be at the end of the search term.

#### Release Note

```release-note
Improved handling of Thai script in search terms
```

### CODE REVIEW

File: server/public/model/search_params.go [hunk 1]
```diff
@@ -11,7 +11,7 @@ import (
 )
 
 var searchTermPuncStart = regexp.MustCompile(`^[^\pL\d\s#"]+`)
-var searchTermPuncEnd = regexp.MustCompile(`[^\pL\d\s*"]+$`)
+var searchTermPuncEnd = regexp.MustCompile(`[^\pL\p{Thai}\d\s*"]+$`)
```
Comment (hmhealey) [inline server/public/model/search_params.go:14]: Now that I'm looking into this, it looks like this regex might remove things like accents in other languages as well. Could we maybe make this a bit broader to also support other languages too?

Looking at the list of Unicode categories on [this website](https://www.regular-expressions.info/unicode.html#prop), it looks like `\p{M}` might work for Thai and other languages as well

**Review by hmhealey — CHANGES_REQUESTED** [review]
Thanks for the PR, @callmeott! I'm a bit surprised to see that `\pL` doesn't handle certain types of letters properly, but I really shouldn't be given how complicated Unicode is.

Would you mind adding a unit test for this so that we can make sure that this won't get broken again in the future?

**Comment by callmeott** [comment]
Hi,

OK let us check if \p{M} works or not for Thai.

Btw, how should we add a unit test? Can we send you a list of Thai search
terms and the expected output after the regex maching?

Cheer,
Pattara

**Comment by callmeott** [comment]
We tested and found that  \p{M}  works fine for Thai.

And here are the example query words that should remain the same after passing this searchTermPuncEnd regex (for unit tests):

สวัสดี
ที่นี่
นี้
กี๊
การันต์
หมาดุ
ลองดู
กู้

Thank you very much,
Pattara

**Comment by hmhealey** [comment]
I think we only need a test or two added to `TestParseSearchFlags2` in `model/search_params_test.go`. You could use anything that didn't work before as the input for it

**Comment by callmeott** [comment]
I've revised the patch to use \p{M} instead and add test cases in
https://github.com/mattermost/mattermost/pull/29241

**Review by hmhealey — APPROVED** [review]
Thanks for adding those tests and for changing that regular expression! This looks great!

**Comment by callmeott** [comment]
Added a related bug fix of Thai search keywords not properly hi-lighted in the search results.

**Comment by isacikgoz** [comment]
@callmeott can you please update the branch, the CI is having a build failure. Thanks!

**Review by isacikgoz — APPROVED** [review]
LGTM, waiting CI to be green.

**Comment by callmeott** [comment]
> @callmeott can you please update the branch, the CI is having a build failure. Thanks!

You mean update the branch at https://github.com/MetamediaTechnology/mattermost right? Done.

**Comment by isacikgoz** [comment]
> > @callmeott can you please update the branch, the CI is having a build failure. Thanks!
> 
> You mean update the branch at https://github.com/MetamediaTechnology/mattermost right? Done.

Yes, if you can also give us the permission to update the branch, we can update before merging it. Thanks!

**Comment by callmeott** [comment]
We noticed that in your latest release, only the changes in server/public/model folder got included but not those in  webapp/channels/src/utils. Is this normal?

**Comment by hmhealey** [comment]
> We noticed that in your latest release, only the changes in server/public/model folder got included but not those in webapp/channels/src/utils. Is this normal?

Which release are you seeing that in? This PR was merged shortly after we started releasing 10.3.0, so nothing from it would've made it into that release, and we only just started the release process for 10.4.0 a few days ago.

If you're comfortable running a release candidate, we did just cut the first one of those for 10.4.0 which would have your changes. The tarballs for the Linux versions of 10.4.0 RC1 are available using these links:
https://releases.mattermost.com/10.4.0-rc1/mattermost-team-10.4.0-rc1-linux-amd64.tar.gz
https://releases.mattermost.com/10.4.0-rc1/mattermost-enterprise-10.4.0-rc1-linux-amd64.tar.gz

**Comment by callmeott** [comment]
We were testing 10.3.0 and found that Thai search was working in some cases so we thought the patch was in. Likely it was rather from DB tuning. 

Sorry for the false alarm, we'll wait for 10.4.0. Thank you very much.

**Pull request merged**